src/lib/support/ThreadOperationalDataset.cpp

```diff
@@ -0,0 +1,525 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <assert.h>
+#include <string.h>
+
+#include "ThreadOperationalDataset.h"
+
+namespace chip {
+namespace Thread {
+
+/**
+ * This class implements Thread Tlv functionality.
+ *
+ */
+class Tlv
+{
+    static constexpr uint8_t kLengthEscape = 0xff; ///< This length value indicates the actual length is of two-bytes length.
+
+public:
+    enum : uint8_t
+    {
+        kChannel         = 0,
+        kPanId           = 1,
+        kExtendedPanId   = 2,
+        kNetworkName     = 3,
+        kPSKc            = 4,
+        kMasterKey       = 5,
+        kMeshLocalPrefix = 7,
+    };
+
+    uint8_t GetSize(void) const { return static_cast<uint8_t>(sizeof(mType) + sizeof(mLength) + GetLength()); }
+
+    /**
+     * This method returns the Tlv type.
+     *
+     * @returns The Tlv type.
+     *
+     */
+    uint8_t GetType(void) const { return mType; }
+
+    /**
+     * This method sets the Tlv type.
+     *
+     */
+    void SetType(uint8_t aType) { mType = aType; }
+
+    /**
+     * This method returns the Tlv length.
+     *
+     * @returns The Tlv length.
+     *
+     */
+    uint8_t GetLength(void) const { return mLength; }
+
+    /**
+     * This method sets the length.
+     */
+    void SetLength(uint8_t aLength) { mLength = aLength; }
+
+    /**
+     * This method returns a pointer to the value.
+     *
+     * @returns The Tlv value.
+     *
+     */
+    const void * GetValue(void) const
+    {
+        return reinterpret_cast<const uint8_t *>(this) + sizeof(mType) +
+            (mLength != kLengthEscape ? sizeof(mLength) : (sizeof(uint16_t) + sizeof(mLength)));
+    }
+
+    void * GetValue(void) { return const_cast<void *>(const_cast<const Tlv *>(this)->GetValue()); }
+
+    /**
+     * This method returns the value as a uint16_t.
+     *
+     * @returns The uint16_t value.
+     *
+     */
+    uint16_t GetValueUInt16(void) const
+    {
+        const uint8_t * p = static_cast<const uint8_t *>(GetValue());
+
+        return static_cast<uint16_t>(p[0] << 8 | p[1]);
```

Comment:
Aren't TLV numbers supposed to be encoded with little-endian? Or it doesn't matter here?

Reply:
Thread TLV forces network-byte order.